### PR TITLE
comp: api: advanced and internal functions separated

### PR DESCRIPTION
**Pull request by mmaka1**
The current content of component.h is a mix of basic api, common
basic helpers for every component developers as well as advanced
functions and macros used by infrastructure and very specialized
components like host, dai, kpb etc.

This patch moves the advanced part to component_ext.h and keeps
only basic part in component.h to avoid overloading effects
component developers with declarations and code they do not use.

Signed-off-by: Marcin Maka <marcin.maka@linux.intel.com>

### CODE REVIEW

File: src/include/sof/audio/component.h [hunk 1]
```diff
@@ -223,70 +222,159 @@ enum comp_copy_type {
 struct comp_driver;
 
 /**
- * Audio component operations - all mandatory.
+ * Audio component operations.
  *
  * All component operations must return 0 for success, negative values for
- * errors and 1 to stop the pipeline walk operation.
+ * errors and 1 to stop the pipeline walk operation unless specified otherwise
+ * in the operation documentation.
```
Comment (paulstelian97) [inline src/include/sof/audio/component.h]: This particular fix in the documentation may warrant a separate commit.

**Comment by mmaka1** [comment]
This is quite a serious change to the layout of component.h but imo very important for effects component developers as greatly reduces scope of api and helpers content they should focus on (sof-docs corresponding change is on its way). I believe that physical separation is better than tons of dox annotations/notes required to highlight the basic and most important content in a single header.

If you know a better name/postfix for component_ext.h please propose.

**Comment by paulstelian97** [comment]
I would suggest the extra documentation on its own be a separate commit.

File: src/include/sof/audio/component.h [hunk 2]
```diff
@@ -47,7 +47,6 @@ struct sof_ipc_stream_posn;
 struct dai_hw_params;
 
 /** \addtogroup component_api Component API
- *  Component API specification.
  *  @{
```
Comment (dbaluta) [inline src/include/sof/audio/component.h]: Why are we removing this line?

File: src/include/sof/audio/component.h [hunk 3]
```diff
@@ -223,70 +222,159 @@ enum comp_copy_type {
...
+	 * @return Pointer to the new component device.
+	 *
+	 * All required data objects should be allocated from the run-time
+	 * heap (RZONE_RUNTIME).
+	 * Any component-specific private data is allocated separately and
+	 * pointer is connected to the comp_dev::private by using
+	 * comp_set_drvdata() and later retrieved by comp_get_drvdata().
+	 *
+	 * All parameters should be initialized to their default values.
+	 */
```
Comment (dbaluta) [inline src/include/sof/audio/component.h]: I would suggest on first factoring out everything that needs to go to component_ext and then enhance documentation. Otherwise, it is hard to understand the intention here.

File: src/include/sof/audio/component.h [hunk 4]
```diff
@@ -303,11 +391,6 @@ struct comp_driver_info {
 	struct list_item list;		/**< list of component drivers */
 };
 
-/** \brief Holds list of registered components' drivers */
-struct comp_driver_list {
-	struct list_item list;	/**< list of component drivers */
-};
-
```
Comment (dbaluta) [inline src/include/sof/audio/component.h]: Why is this removed from here. I would think that this is a basic structure that needs to be here.

File: src/include/sof/audio/component.h (hunk 2 above)
Comment (mmaka1) [reply src/include/sof/audio/component.h]: When you read the header, it is just repeated *Component API* from the line above, no value here. Breathe also does not use this additional text in any way. The group is just rendered inline in the place of reference, so it does not appear as longer former of "Component API" anywhere.

File: src/include/sof/audio/component.h (hunk 4 above)
Comment (mmaka1) [reply src/include/sof/audio/component.h]: This structure is used by the infrastructure only, in a single place. Component developers do not need to know it exists imo.

File: src/include/sof/audio/component.h (hunk 1 above)
Comment (mmaka1) [reply src/include/sof/audio/component.h]: ok

File: src/include/sof/audio/component.h (hunk 2 above)
Comment (dbaluta) [reply src/include/sof/audio/component.h]: @mmaka1 this is the kind of information we need in the commit message. sorry for troubling you with this.

File: src/include/sof/audio/component.h (hunk 4 above)
Comment (dbaluta) [reply src/include/sof/audio/component.h]: Ok.

File: src/include/sof/audio/component.h (hunk 3 above)
Comment (mmaka1) [reply src/include/sof/audio/component.h]: Moved dox related changes to a separate patch.
cc @paulstelian97
